drivers/greth.c

```diff
@@ -0,0 +1,564 @@
+}
+
+
+static const mdio_bus_ops_t greth_mdio_ops = {
+	greth_mdioSetup,
+	greth_mdioRead,
+	greth_mdioWrite,
+};
+
+
+static void greth_setLinkState(void *arg, int state)
+{
+	struct netif *netif = arg;
+	greth_state_t *s = netif->state;
+	int speed, full_duplex;
+	if (state != 0) {
+		speed = ephy_linkSpeed(&s->phy, &full_duplex);
+		if (speed == 10) {
+			uint32_t ctr = s->mmio->CTRL;
+			ctr &= ~(full_duplex ? 0 : GRETH_CTRL_FD);
+			ctr &= ~(GRETH_CTRL_SP);
+			ctr &= ~(GRETH_CTRL_GB);
+			s->mmio->CTRL = ctr;
+		}
+		if (speed == 100) {
+			/* GRETH_CTRL_SP = 1 when 100 Mbit and 0 otherwise */
+			uint32_t ctr = s->mmio->CTRL;
+			ctr &= ~(full_duplex ? 0 : GRETH_CTRL_FD);
+			ctr |= GRETH_CTRL_SP;
+			ctr &= ~(GRETH_CTRL_GB);
+			s->mmio->CTRL = ctr;
+		}
+		if (speed == 1000) {
+			/* Half duplex not supported in Gbit mode */
+			uint32_t ctr = s->mmio->CTRL;
+			ctr |= GRETH_CTRL_FD;
+			ctr &= ~(GRETH_CTRL_SP);
+			ctr |= GRETH_CTRL_GB;
+			s->mmio->CTRL = ctr;
+		}
+		netif_set_link_up(netif);
+	}
+	else {
+		netif_set_link_down(netif);
+	}
+}
+
+/* TX / RX */
+static size_t greth_nextRxBufferSize(const net_bufdesc_ring_t *ring, size_t i)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+	size_t sz;
+	if ((desc->flags & GRETH_DESC_EN) != 0) {
+		return 0;
+	}
+	sz = (desc->flags >> GRETH_DESC_LEN_SHIFT) & GRETH_DESC_LEN_MASK;
+	if (sz == 0) {
+		sz = 1;
+		printf("lwip: greth: WARNING: This message indicates a potential HW bug:\n");
+		printf("lwip: greth: HW provided invalid size: %zu. Setting size to 1\n", sz);
+	}
+	return sz;
+}
+
+
+static int greth_pktRxFinished(const net_bufdesc_ring_t *ring, size_t i)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+
+	return !(desc->flags & GRETH_DESC_EN);
+}
+
+
+static void greth_fillRxDesc(const net_bufdesc_ring_t *ring, size_t i, addr_t pa, size_t sz, unsigned seg)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+	unsigned wrap = desc == (volatile greth_buf_desc_t *)ring->ring + ring->last ? GRETH_DESC_WR : 0;
+
+	desc->addr = pa;
+
+	atomic_thread_fence(memory_order_seq_cst);
+	desc->flags = GRETH_DESC_EN | GRETH_DESC_IE | wrap;
+	atomic_thread_fence(memory_order_seq_cst);
+}
+
+
+static int greth_nextTxDone(const net_bufdesc_ring_t *ring, size_t i)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+	return !(desc->flags & GRETH_DESC_EN);
+}
+
+
+static void greth_fillTxDesc(const net_bufdesc_ring_t *ring, size_t i, addr_t pa, size_t sz, unsigned seg)
+{
+	volatile greth_buf_desc_t *desc = (volatile greth_buf_desc_t *)ring->ring + i;
+	uint32_t flags;
+
+	flags = GRETH_DESC_EN;
+	flags |= (i == ring->last) ? GRETH_DESC_WR : 0;
+	flags |= (sz << GRETH_DESC_LEN_SHIFT) & GRETH_DESC_LEN_MASK; /* shift (0) just for consistency  */
+	flags |= ((seg & BDRING_SEG_LAST) != 0) ? 0 : GRETH_DESC_TX_MO;
+	flags |= GRETH_DESC_IE;
+
+#if GRETH_CHECKSUM_OFFLOAD
```

Comment:
Is there a situation in which we'd not want to offload it? If not, then I'd remove the `#define`, otherwise please move it to `board_config.h`